### PR TITLE
feat: Personal Access Tokens

### LINKED ISSUE #364: support for PAT in hasura-auth
# Motivation

Today when a user needs to interact with a hasura-auth/hasura backed service outside of a frontend application they have two options:

1. Share the admin secret
2. Implement something customized (like we did ourselves for the cli)

# Goal

The goal is to be able to generate tokens to authenticate a backend service or tool as a user. For instance, a `cli` tool, or some backend service that needs access to hasura.

# Potential implementation

A potential implementation could to simply rely on the existing refresh-tokens functionality. In this scenario, a tool like the cli could:

1. Authenticate the user as normally done
2. Have the user request a PAT
3. Store it

At this point the CLI could just reuse that refresh token and act on behalf of the user.

A similar workflow could be used to authenticate backend services:

1. An admin could create a "service account"
2. Give it a role that has permissions to do what the backend service needs to do
3. Create a PAT/refresh_token
4. Configure it in the backend service

## Requirements

For this to work I guess we need:

1. A mechanism to generate refresh tokens with a longer timespan, i.e., 1 year, 2 years, 99 years....
2. A mechanism to list them and invalidate them

## Security implications

The benefit of this mechanism is that the PAT needs to be authenticated every time to generate an access token so it can be revoked anytime.

### PR DESCRIPTION
**Pull request by szilarddoro**
Resolves #364

### CODE REVIEW

File: migrations/00013_add_refresh_token_type.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+BEGIN;
```
Comment (dbarrosop) [inline migrations/00013_add_refresh_token_type.sql:1]: does this work with hasura?

Comment (dbarrosop) [reply migrations/00013_add_refresh_token_type.sql:1]: it actually works, this must be new. They used to suggest using a different table for this. Nice!

File: src/routes/pat/pat.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { sendError } from '@/errors';
+import { getUser, gqlSdk } from '@/utils';
+import { RequestHandler } from 'express';
+import Joi from 'joi';
+import { v4 as uuidv4 } from 'uuid';
+
+export const createPATSchema = Joi.object({
+  expiresAt: Joi.date().required(),
+  metadata: Joi.object(),
```
Comment (dbarrosop) [inline src/routes/pat/pat.ts:9]: what is this metadata?

Comment (dbarrosop) [reply src/routes/pat/pat.ts:9]: ignore me

File: src/routes/pat/pat.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}).meta({ className: 'CreatePATSchema' });
+
+export const createPATHandler: RequestHandler<
+  {},
+  {},
+  { refreshToken: string; metadata: object; expiresAt: Date }
```
Comment (dbarrosop) [inline src/routes/pat/pat.ts:15]: shouldn't `refreshToken` be an `accessToken`? In which case you'd just validate the signature and create a PAT for the user specified in the JWT

**Comment by dbarrosop** [comment]
The workflow looks good, should we provide a mechanism for:

- getting PATs
- revoking/deleting PATs

Or is the expectation people will use graphql queries for that? I think that's it, right?

**Review by dbarrosop — APPROVED** [review]

File: src/routes/pat/pat.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  if (!req.auth) {
+    return sendError(res, 'unauthenticated-user');
+  }
+
+  const { userId } = req.auth as RequestAuth;
+
+  const user = await getUser({ userId });
+
+  if (!user) {
+    return sendError(res, 'invalid-refresh-token');
```
Comment (nunopato) [inline src/routes/pat/pat.ts:26]: did you mean user instead of refresh token here?

File: src/routes/pat/pat.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  const { metadata, expiresAt } = req.body;
+
+  // Note: Joi wouldn't work here because we need to compare the date to the
+  // date of the request, not the date when the schema was created
+  // 7 days
```
Comment (nunopato) [inline src/routes/pat/pat.ts:33]: out of curiosity, what is the reasoning for 7 days?

Comment (szilarddoro) [reply src/routes/pat/pat.ts:33]: I followed GitHub's approach in this. The minimum amount of expiration for their legacy tokens is 7 days. We can change it to anything.

File: src/routes/pat/pat.ts (hunk 4 above)
Comment (szilarddoro) [reply src/routes/pat/pat.ts:26]: Good catch. Will fix this.

File: src/routes/signin/pat.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+export const signInPATSchema = Joi.object({
+  personalAccessToken,
+}).meta({ className: 'SignInPATSchema' });
+
+export const signInPATHandler: RequestHandler<
+  {},
+  {},
+  { personalAccessToken: string }
+> = async (req, res) => {
+  const user = await getUserByRefreshToken(req.body.personalAccessToken);
```
Comment (dbarrosop) [inline src/routes/signin/pat.ts:16]: should we also validate the type of the refresh token is `PAT`?

Comment (szilarddoro) [reply src/routes/signin/pat.ts:16]: Yeah, good point. I'll create a separate GraphQL query for this one.

File: src/routes/pat/pat.ts (hunk 5 above)
Comment (nunopato) [reply src/routes/pat/pat.ts:33]: :+1:

**Pull request merged**